sys/kern/proc.c

```diff
@@ -339,48 +339,75 @@ static bool is_zombie(proc_t *p) {
 }
 
-/* Wait for direct children. */
+static bool child_need_attention(proc_t *child, pid_t pid, pgrp_t *pg) {
+  /* pid > 0 => child with PID same as pid */
+  if (pid == child->p_pid)
+    return true;
+  /* pid == -1 => any child  */
+  if (pid == -1)
+    return true;
+  /* pid == 0 => child with PGID same as ours */
+  if (pid == 0 && child->p_pgrp == pg)
+    return true;
+  /* pid < -1 => child with PGID equal to -pid */
+  if (pid < -1 && child->p_pgrp->pg_id != -pid)
+    return true;
+  return false;
+}
+
+/* Wait for direct children.
+ * Pointers to output parameters must point to valid kernel memory. */
 int do_waitpid(pid_t pid, int *status, int options, pid_t *cldpidp) {
   proc_t *p = proc_self();
 
   WITH_MTX_LOCK (all_proc_mtx) {
-    /* Start with zombies, if no zombies wait for a child to become one. */
     for (;;) {
-      proc_t *child = NULL;
+      bool any = false;
+      proc_t *child;
 
       /* Check children meeting criteria implied by pid. */
       TAILQ_FOREACH (child, CHILDREN(p), p_child) {
-        /* pid > 0 => child with PID same as pid */
-        if (pid == child->p_pid)
-          break;
-        /* Lookup zombie children */
+        if (!child_need_attention(child, pid, p->p_pgrp))
+          continue;
+
+        any = true;
+
+        /*
+         * It's not necessary to lock the child here, since:
+         * a) We're holding all_proc_mtx, so it won't get deleted while
+         *    we're inspecting it;
+         * b) We're only doing unprotected atomic reads of p_state.
+         */
+
+        *cldpidp = child->p_pid;
+
         if (is_zombie(child)) {
-          /* pid == -1 => any child  */
-          if (pid == -1)
-            break;
-          /* pid == 0 => child with PGID same as ours */
-          if ((pid == 0) && (child->p_pgrp == p->p_pgrp))
-            break;
-          /* pid < -1 => child with PGID equal to -pid */
-          if (pid < -1 && (child->p_pgrp->pg_id != -pid))
-            break;
+          *status = child->p_exitstatus;
+          proc_reap(child);
+          return 0;
         }
-      }
 
-      /* No child with such pid. */
-      if (!child && pid > 0)
-        return ECHILD;
+        if ((options & WUNTRACED) && (child->p_state == PS_STOPPED) &&
+            (child->p_flags & PF_STOPPED)) {
+          child->p_flags &= ~PF_STOPPED;
+          *status = MAKE_STATUS_SIG_STOP(SIGSTOP);
+          return 0;
+        }
 
-      if (child && is_zombie(child)) {
-        if (status)
-          *status = child->p_exitstatus;
-        pid_t pid = child->p_pid;
-        proc_reap(child);
-        *cldpidp = pid;
-        return 0;
+        if ((options & WCONTINUED) && (child->p_state == PS_NORMAL) &&
+            (child->p_flags & PF_CONTINUED)) {
+          child->p_flags &= ~PF_CONTINUED;
+          *status = MAKE_STATUS_SIG_CONT();
+          return 0;
+        }
+
+        if (pid > 0)
+          break;
       }
 
-      /* No zombie child was found. */
+      if (!any)
```

Comment:
Is there a way to check the same thing without introducing boolean variable?

Reply:
I couldn't think of any obvious way to do it.

Reply:
I guess we could introduce an `error` variable like the one here: http://bxr.su/NetBSD/sys/kern/kern_exit.c#1021

Reply:
Done!